satellite/satellitedb/overlaycache.go

```diff
@@ -987,6 +987,113 @@ func (cache *overlaycache) UpdatePieceCounts(ctx context.Context, pieceCounts ma
 	return Error.Wrap(err)
 }
 
+// GetExitingNodes returns nodes who have initiated a graceful exit.
+func (cache *overlaycache) GetExitingNodes(ctx context.Context) (exitingNodes storj.NodeIDList, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	rows, err := cache.db.Query(cache.db.Rebind(`
+		SELECT id FROM nodes
+		WHERE exit_initiated_at IS NOT NULL
+		AND exit_finished_at IS NULL
+		`),
+	)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		err = errs.Combine(err, rows.Close())
+	}()
+
+	for rows.Next() {
+		var id storj.NodeID
+		err = rows.Scan(&id)
+		if err != nil {
+			return nil, err
+		}
+		exitingNodes = append(exitingNodes, id)
+	}
+	return exitingNodes, nil
+}
+
+// GetExitingNodesLoopIncomplete returns exiting nodes who haven't completed the metainfo loop iteration.
+func (cache *overlaycache) GetExitingNodesLoopIncomplete(ctx context.Context) (exitingNodes storj.NodeIDList, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	rows, err := cache.db.Query(cache.db.Rebind(`
+		SELECT id FROM nodes
+		WHERE exit_initiated_at IS NOT NULL
+		AND exit_loop_completed_at IS NULL
+		AND exit_finished_at IS NULL
```

Comment:
In the initial release, "`exit_finished_at` will always be null when `exit_loop_completed_at` is null" should be true.  However, we also discussed a future feature that would allow the SNO to cancel the graceful exit. Albeit, not likely it would happen before the metainfo loop completed.

Reply:
I agree that we should account for the possibility that `exit_loop_completed_at` is NULL while `exit_finished_at` is not. A test case has also been added for this.

Reply:
If it could happen, my main concern is whether such rows ought to be considered "loop incomplete" and included in the results from this function. It kind of seems like they shouldn't, but I haven't studied all of the details of the behavior.

Reply:
I think @ethanadams might know better, but I can't see a situation where we would want to return "loop incomplete" nodes that have already finished exiting. Off the top of my head I don't know what this function is for, but I would guess that whatever calls it doesn't care about fully exited nodes.